Xplat/src/main/java/vazkii/botania/test/item/AstrolabeTest.java

```diff
@@ -0,0 +1,405 @@
+		final int expectedMana = (MANA_BARELY_ENOUGH - COST_ASTROLABE) % DepthsRodItem.COST;
+		checkRemainingMana(player, expectedMana);
+
+		final int expectedNumberOfBlocks = (MANA_BARELY_ENOUGH - COST_ASTROLABE) / DepthsRodItem.COST;
+		final MutableInt actualNumberOfBlocks = new MutableInt();
+		helper.forEveryBlockInStructure(blockPos -> {
+			if (helper.getBlockState(blockPos).is(Blocks.COBBLESTONE)) {
+				actualNumberOfBlocks.increment();
+			}
+		});
+		TestingUtil.assertEquals(expectedNumberOfBlocks, actualNumberOfBlocks.intValue(),
+				() -> String.format("Expected %d placed blocks, found %d",
+						expectedNumberOfBlocks, actualNumberOfBlocks.intValue()));
+
+		helper.succeed();
+	}
+
+	private Player mockAstrolabeAndUseWithCobbleRod(GameTestHelper helper, int mana) {
+		final Player player = mockPlayerWithAstrolabe(helper, POS_PLAYER_MANA, POS_PLAYER_MANA, mana);
+		player.addItem(new ItemStack(BotaniaItems.cobbleRod));
+		final ItemStack stack = getAstrolabeForBlockType(player, Blocks.COBBLESTONE);
+		ItemNBTHelper.setInt(stack, AstrolabeItem.TAG_SIZE, SIZE_ASTROLABE);
+
+		useAstrolabe(player, stack, InteractionHand.MAIN_HAND);
+
+		return player;
+	}
+	//endregion
+
+	private static final int SLOT_TABLET = 1;
+
+	private void useAstrolabe(Player player, ItemStack stack, InteractionHand hand) {
+		BlockHitResult hitResult = ToolCommons.raytraceFromEntity(player, 5, true);
+		UseOnContext useOnContext = new UseOnContext(player, hand, hitResult);
+		stack.useOn(useOnContext);
+	}
+
+	private void checkRemainingMana(Player player, int expectedMana) {
+		final ItemStack tablet = player.getInventory().getItem(SLOT_TABLET);
+		final ManaItem manaItem = XplatAbstractions.INSTANCE.findManaItem(tablet);
+		TestingUtil.assertThat(manaItem != null, () -> "Missing mana tablet");
+		TestingUtil.assertEquals(expectedMana, manaItem.getMana(),
+				() -> String.format("Expected %d remaining mana, but found %d", expectedMana, manaItem.getMana()));
+	}
+
+	@NotNull
+	private Player mockPlayerWithAstrolabe(final GameTestHelper helper, Vec3 posPlayer, final Vec3 lookTarget, int mana) {
+		final Player player = helper.makeMockPlayer();
+		player.setItemInHand(InteractionHand.MAIN_HAND, new ItemStack(BotaniaItems.astrolabe));
+		final ItemStack tablet = new ItemStack(BotaniaItems.manaTablet);
+		Objects.requireNonNull(XplatAbstractions.INSTANCE.findManaItem(tablet)).addMana(mana);
+		player.getInventory().add(tablet);
+		player.setPos(helper.absoluteVec(posPlayer));
+		final Vec3 targetVec = helper.absoluteVec(lookTarget);
+		player.lookAt(EntityAnchorArgument.Anchor.EYES, targetVec);
+		return player;
+	}
+
+	@NotNull
+	private ItemStack getAstrolabeForBlockType(final Player player, Block block) {
+		final ItemStack stack = player.getItemInHand(InteractionHand.MAIN_HAND);
+		ItemNBTHelper.setCompound(stack, AstrolabeItem.TAG_BLOCKSTATE,
+				NbtUtils.writeBlockState(block.defaultBlockState()));
```

Comment:
expose and use util method from AstrolabeItem instead of assuming the save format in the test?